extensions/s3/src/main/java/com/hazelcast/jet/s3/S3Sources.java

```diff
@@ -179,13 +237,32 @@ private S3Sources() {
                 .build();
     }
 
+    /**
+     * Functional interface for functions reading some file with context information of key and bucket name.
+     *
+     * @param <I> Type of elements returned from reading.
+     */
+    @FunctionalInterface
+    public interface FileReadingFunction<I> extends Serializable {
```

Comment:
we can make use of `com.hazelcast.jet.function.TriFunction<T0, T1, T2, R>` instead of introducing a functional interface here